### PR TITLE
[resharding] Delete snapshot that was created for resharding in the next epoch

### PR DESCRIPTION
**Pull request by shreyan-gupta**
This is a follow up to the original PR https://github.com/near/nearcore/pull/9967

This PR adds a new delete_snapshot_callback that chain can call to ONLY delete snapshot and not create a new one.

This is useful for when we have the snapshot_config_type as ForReshardingOnly. We would ideally like to delete the snapshot at the beginning of the next epoch once the resharding is completed.

### CODE REVIEW

File: chain/chain/src/state_snapshot_actor.rs [hunk 1]
```diff
@@ -144,3 +151,16 @@ pub fn get_make_snapshot_callback(
...
+
+/// Sends a request to delete a state snapshot.
+pub fn get_delete_snapshot_callback(
+    state_snapshot_addr: Arc<actix::Addr<StateSnapshotActor>>,
+) -> DeleteSnapshotCallback {
+    Arc::new(move || {
+        tracing::info!(
+            target: "state_snapshot",
+            "start_delete_snapshot_callback sends `DeleteSnapshotRequest` to state_snapshot_addr");
+        state_snapshot_addr
```
Comment (shreyan-gupta) [inline chain/chain/src/state_snapshot_actor.rs:163]: We don't need to call `flat_storage_manager.set_flat_state_updates_mode(false)` here

File: integration-tests/src/tests/client/resharding.rs [hunk 2]
```diff
@@ -605,6 +606,57 @@ impl TestReshardingEnv {
             }
         }
     }
+
+    // function to check whether the state snapshot exists or not and whether it exists at the correct
+    // hash for both cases when state snapshot is enabled and disabled
+    fn check_snapshot(&mut self, state_snapshot_enabled: bool) {
```
Comment (shreyan-gupta) [inline integration-tests/src/tests/client/resharding.rs:612]: This is the same test from https://github.com/near/nearcore/pull/9967

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -4321,11 +4328,15 @@ impl Chain {
             }
         };
 
-        if need_snapshot {
+        // We need to delete the existing snapshot at the epoch boundary if we are not making a new snapshot
+        // This is useful for the next epoch after resharding where make_snapshot is false but it's an epoch boundary
+        let delete_snapshot = !make_snapshot && is_epoch_boundary;
```
Comment (nikurt) [inline chain/chain/src/chain.rs:4333]: Very nice!

**Review by nikurt — APPROVED** [review]

**Comment by wacban** [comment]
> We would ideally like to delete the snapshot at the beginning of the next epoch once the resharding is completed.

Ideally we would delete the snapshot as soon as resharding is finished. The snapshot overhead grows over the duration of epoch and full epoch worth of overhead may be too much on archival nodes. That being said this PR is a good first step in the right direction.

**Review by wacban — APPROVED** [review]
LGTM

**Comment by shreyan-gupta** [comment]
> > We would ideally like to delete the snapshot at the beginning of the next epoch once the resharding is completed.
> 
> Ideally we would delete the snapshot as soon as resharding is finished. The snapshot overhead grows over the duration of epoch and full epoch worth of overhead may be too much on archival nodes. That being said this PR is a good first step in the right direction.

We now have the flexibility of having this delete_snapshot_callback. What we can later do is that in the resharding post processing step (which is called after resharding finishes), check for appropriate configs and delete snapshot.

**Pull request merged**